torchdynamo/symbolic_convert.py

```diff
@@ -490,31 +508,40 @@ def IMPORT_NAME(self, inst):
         fromlist = fromlist.as_python_constant()
         module_name = inst.argval
 
-        value = __import__(
-            module_name,
-            fromlist=fromlist,
-            level=level,
-            globals=self.f_globals,
-        )
-
-        if level != 0:
-            pkg = self.calc_package()
-            module_name = self.resolve_name(module_name, pkg, level)
-
-        # For __import__, when the name variable is of the form package.module,
-        # normally, the top-level package (the name up till the first dot) is
-        # returned, not the module named by module_name. However, when a
-        # non-empty fromlist argument is given, the module named by name is
-        # returned. Therefore, we set the source correctly here.
-        if not fromlist:
-            top_level_module_name = module_name.partition(".")[0]
-            source = self.import_source(top_level_module_name)
+        # Are we replaying? if so, load recorded module
+        recorded_name = ExecutionRecorder.LOCAL_MOD_PREFIX + module_name
+        if recorded_name in self.f_globals:
+            value = self.f_globals[recorded_name]
+            source = GlobalSource(recorded_name)
         else:
-            source = self.import_source(module_name)
+            value = __import__(
+                module_name,
+                fromlist=fromlist,
+                level=level,
+                globals=self.f_globals,
+            )
+
+            if level != 0:
+                pkg = self.calc_package()
+                module_name = self.resolve_name(module_name, pkg, level)
+
+            # For __import__, when the name variable is of the form package.module,
+            # normally, the top-level package (the name up till the first dot) is
+            # returned, not the module named by module_name. However, when a
+            # non-empty fromlist argument is given, the module named by name is
+            # returned. Therefore, we set the source correctly here.
+            if not fromlist:
+                top_level_module_name = module_name.partition(".")[0]
+                source = self.import_source(top_level_module_name)
+            else:
+                source = self.import_source(module_name)
+
+        if config.replay_record_enabled:
+            self.exec_recorder.add_local_mod(module_name, value)
 
         if is_allowed(value):
             self.push(TorchVariable(value, source=source))
-        elif istype(value, types.ModuleType):
+        elif istype(value, (types.ModuleType, DummyModule)):
```

Comment:
Can't we make `DummyModule` a subclass of `types.ModuleType` so this isn't needed everywhere?

Reply:
This was my first approach, but it breaks if we do this, because `dill` tries to import it.